File: sortedSquares.cpp
//给你一个按 非递减顺序 排序的整数数组 nums，返回 每个数字的平方 组成的新数组，要求也按 非递减顺序 排序。
//
//
//
//
//
//
// 示例 1：
//
//
//输入：nums = [-4,-1,0,3,10]
//输出：[0,1,9,16,100]
//解释：平方后，数组变为 [16,1,0,9,100]
//排序后，数组变为 [0,1,9,16,100]
//
// 示例 2：
//
//
//输入：nums = [-7,-3,2,3,11]
//输出：[4,9,9,49,121]
//
//
//
//
// 提示：
//
//
// 1 <= nums.length <= 10⁴
// -10⁴ <= nums[i] <= 10⁴
// nums 已按 非递减顺序 排序
//
//
//
//
// 进阶：
//
//
// 请你设计时间复杂度为 O(n) 的算法解决本问题
//
// Related Topics 数组 双指针 排序 👍 346 👎 0
using namespace std;

#include "vector"

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    vector<int> sortedSquares(vector<int> &nums) {
        // 双指针
        // loop vector once, time complexity O(n)
        // brute force: square everyone O(n) + sorting O(nlogn) = O(n+nlogn)
        int i = 0, j = nums.size() - 1, k = nums.size() - 1;
        vector<int> ans(nums.size());

        while (i <= j) {
            if (nums[i] * nums[i] > nums[j] * nums[j]) {
                ans[k--] = nums[i] * nums[i];
                i++;
            } else {
                ans[k--] = nums[j] * nums[j];
                j--;
            }
        }
        return ans;
    }
};

//leetcode submit region end(Prohibit modification and deletion)
int main() {
    Solution *solution = new Solution();
    vector<int> nums = {-7, -3, 2, 3, 11};
    vector<int> answer = solution->sortedSquares(nums);
    return 0;
}

File: sortedSquares.py
# Given an integer array nums sorted in non-decreasing order, return an array
# of the squares of each number sorted in non-decreasing order.
#
#
#  Example 1:
#
#
# Input: nums = [-4,-1,0,3,10]
# Output: [0,1,9,16,100]
# Explanation: After squaring, the array becomes [16,1,0,9,100].
# After sorting, it becomes [0,1,9,16,100].
#
#
#  Example 2:
#
#
# Input: nums = [-7,-3,2,3,11]
# Output: [4,9,9,49,121]
#
#
#
#  Constraints:
#
#
#  1 <= nums.length <= 10⁴
#  -10⁴ <= nums[i] <= 10⁴
#  nums is sorted in non-decreasing order.
#
#
#
# Follow up: Squaring each element and sorting the new array is very trivial,
# could you find an
# O(n) solution using a different approach?
#
#  Related Topics Array Two Pointers Sorting 👍 8470 👎 211

from typing import List

# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        l,r = 0, len(nums)-1
        # nums is sorted
        # start find the largest number, then coming close to the center
        res = []
        # building in resverse order b/c we add the large result in first
        while l<=r:
            # since the absolute value of the number will be smaller when the number in the middle(list is sorted),
            # we start from the middle, then use ptr spread towards left & right
            if nums[l] ** 2 > nums[r] ** 2:
                res.append(nums[l]**2)
                l+=1
            else:
                res.append(nums[r]**2)
                r-=1
        res.reverse()
        return res
# time : O(logn)
# space: O(n)

# leetcode submit region end(Prohibit modification and deletion)

if __name__ == "__main__":
    nums = [-7, -3, 2, 3, 11]
    s = Solution()
    print(s.sortedSquares(nums))